Canonical kw detection

ruohanxiong opened the pull request:
Adds ability for researchers to determine canonical (minimum) search query keywords that trigger censorship. Implemented by recursively bisecting the search query to determine block(s) triggering censorship, then iterating through the shortened query removing one character at a time. Characters that when removed, cause censorship message to no longer show, are considered part of canonical censorship keyword. 

Enabled via parameter `get_canonical=True` (default=False) passed to `run()`. Helper function `split_search_query()` separately for recursive bisection. Changes have been tested.

Database table is modified to accommodate two new columns (`is_canonical bool` and `orig_keyword string`). Existing databases will need to be modified. A database updater script `update_db.py` is included.

README has been updated to include info on canonical keyword detection, migrating database, and tables in README now reflect new database table.  

Increments blocked-on-weibo project to v0.2

jasonqng commented on blockedonweibo/weibo.py:
```diff
@@ -25,16 +25,17 @@
-from . import weibo_credentials
+try:
+    from . import weibo_credentials
+except ValueError:
```
Good catch, thank you!

jasonqng commented on blockedonweibo/weibo.py:
```diff
@@ -319,17 +328,52 @@ def run(keywords,
                                     pd.DataFrame([{"date":date,
                                                    "datetime":datetime.now(),
-                                                   "keyword":r.keyword,
+                                                   "keyword":min_str if min_str is not None else r.keyword,
```
Looks like min_str isn't in scope. Maybe move the declare on line 342 to up before the loop? Perhaps that'll fix it? (Sorry, I should probably write tests at some point to save folks time with catching stuff like this.)

```
>>> import weibo
>>> import pandas as pd
>>> sqlite_file = 'results.sqlite'
>>> weibo.create_database(sqlite_file, overwrite=True)
>>> sample_keywords_df = pd.DataFrame(
...     [{'keyword':'hello','source':'my dataframe'},
...      {'keyword':'lxb','source':'my dataframe'},
...      {'keyword':u'习胞子','source':'my dataframe'}
...     ])
>>> weibo.run(sample_keywords_df,insert=False,return_df=True)
(0, u'hello', 'has_results')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "weibo.py", line 370, in run
    "keyword":min_str if min_str is not None else r.keyword,
UnboundLocalError: local variable 'min_str' referenced before assignment
```

ruohanxiong replied:
Whoa.. good catch, thank you! Corrected by declaring min_str before loop

jasonqng commented:
@ruohanxiong Thanks so much for adding this in! I've been meaning to do it, but this is so much more elegant than anything I'd do.

Also appreciate your cleaning up the code. Eyeballing it, it looks solid, but can you fix the min_str issue and make sure it can run with the test cases in the readme? I'll test again once you get it working and then we can merge (I'll add writing tests to the backlog).

Only other suggestion might be adding a short docstring inside your new update file, but if you're busy, no worries.

Thanks again for the contribution!

jasonqng commented on blockedonweibo/weibo.py:
```diff
@@ -338,4 +382,24 @@ def run(keywords,
+
+
+def split_search_query(query, cookies, sleep=0, res_rtn=[], known_blocked=False):
+    """Recursively halves a query and returns portions with blocked keywords as a list of strings.
```
Can you add to the docstring a short explanation for res_rtn and known_blocked (e.g., what's the purpose of known_blocked, doesn't look like it's being used; for future extensibility/skipping if it's already detected as censored?)

ruohanxiong commented:
@jasonqng 
Good catch on the min_str declaration. I've updated branch to correct, and added docstrings to update_db.py script. Added notes to docstring of the added recursive helper function (split_search_query in weibo.py) also.  

Thanks for reviewing and for the comments! Was a pleasure working on blockedonweibo - the project was structured beautifully.

jasonqng commented on blockedonweibo/weibo.py:
```diff
@@ -319,17 +328,53 @@ def run(keywords,
+                    if verbose=="all":
+                        print("Testing %d of %d: omitting character %s" %(i+1, len(test_list), kw[i]))
+                    if sleep_recursive:
+                        time.sleep(random.randint(math.ceil(sleep * .90), math.ceil(sleep * 1.10)))
```
sleep_secs, not sleep

jasonqng commented:
@ruohanxiong Can you confirm that the below is expected? It doesn't seem to do splitting of the term, but only removes one character at a time. Also, the keyword and orig_keyword don't seem to match the sample_keywords_df?

```
>>> import weibo
>>> import pandas as pd
>>> sqlite_file = 'results.sqlite'
>>> weibo.create_database(sqlite_file, overwrite=True)
>>> cookie = weibo.load_cookies()
>>> print(weibo.verify_cookies_work(cookie))
True
>>> sample_keywords_df = pd.DataFrame(
...     [{'keyword':'hello','source':'my dataframe'},
...         {'keyword':'lxb ok','source':'my dataframe'},
...         {'keyword':'yes lxb ok','source':'my dataframe'},
...         {'keyword':'nolxbok','source':'my dataframe'},
...         {'keyword':'lxb','source':'my dataframe'},
...         {'keyword':u'习胞子','source':'my dataframe'}
...     ])
>>> weibo.run(sample_keywords_df,insert=True,return_df=True,get_canonical=True,sqlite_file=sqlite_file,cookies=cookie)
(0, u'hello', 'has_results')
(1, u'lxb ok', 'censored')
Found censored search phrase; determining canonical censored keyword set
[u'lxb']
Testing 1 of 3: omitting character l
Testing 2 of 3: omitting character x
Testing 3 of 3: omitting character b
the minimum phrase from 'lxb' is: 'lxb'
(3, u'nolxbok', 'censored')
Found censored search phrase; determining canonical censored keyword set
[u'nolxbok']
Testing 1 of 7: omitting character n
Testing 2 of 7: omitting character o
Testing 3 of 7: omitting character l
Testing 4 of 7: omitting character x
Testing 5 of 7: omitting character b
Testing 6 of 7: omitting character o
Testing 7 of 7: omitting character k
the minimum phrase from 'nolxbok' is: 'lxb'
/Users/jasonng/anaconda2/lib/python2.7/site-packages/pandas/core/algorithms.py:410: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
  f = lambda x, y: htable.ismember_object(x, values)
(5, u'\u4e60\u80de\u5b50', 'no_results')
         date                   datetime  is_canonical keyword num_results  \
0  2017-11-30 2017-11-30 16:17:15.370630         False   hello    76838069
0  2017-11-30 2017-11-30 16:19:30.766213          True     lxb        None
0  2017-11-30 2017-11-30 16:22:41.568311          True     lxb        None
0  2017-11-30 2017-11-30 16:22:57.953655         False     习胞子        None

  orig_keyword       result        source  test_number
0         None  has_results  my dataframe            1
0       lxb ok     censored  my dataframe            1
0      nolxbok     censored  my dataframe            1
0         None   no_results  my dataframe            1
>>> weibo.sqlite_to_df(sqlite_file)
   id        date             datetime_logged  test_number  keyword  censored  \
0   0  2017-11-30  2017-11-30 16:17:15.368423            1    hello         0
1   1  2017-11-30  2017-11-30 16:19:30.761193            1      lxb         0
2   2  2017-11-30  2017-11-30 16:19:30.765220            1   lxb ok         0
3   1  2017-11-30  2017-11-30 16:22:41.561827            1      lxb         0
4   4  2017-11-30  2017-11-30 16:22:41.567304            1  nolxbok         0
5   5  2017-11-30  2017-11-30 16:22:57.951499            1      习胞子         0

   no_results  reset  is_canonical       result        source orig_keyword  \
0           0      0             0  has_results  my dataframe         None
1           0      0             1     censored  my dataframe       lxb ok
2           0      0             0     censored  my dataframe         None
3           0      0             1     censored  my dataframe      nolxbok
4           0      0             0     censored  my dataframe         None
5           0      0             0   no_results  my dataframe         None

   num_results notes
0   76838069.0  None
1          NaN  None
2          NaN  None
3          NaN  None
4          NaN  None
5          NaN  None
```

ruohanxiong commented:
@jasonqng Thanks - corrected sleep to sleep_secs.

Traced through your input example below. Can confirm that the recursive function works as intended, though there was another important bug that your test caught here. 

Canonical keyword is actually detected by the character omission loop but the recursive function helps by continuously halving the query (stopping at the step before no more censorship occurs) so that uncensored parts of the query are removed for better performance. 

Example (works better with longer queries)

```
from blockedonweibo import weibo
import pandas as pd
cookie = weibo.load_cookies()
print(weibo.verify_cookies_work(cookie))
True
weibo.run(['lxb the quick brown fox '],insert=False,return_df=True,get_canonical=True,cookies=cookie)

(0, u'lxb the quick brown fox ', 'censored')
Found censored search phrase; determining canonical censored keyword set
Recursively shortening... testing query: "lxb the quic"
Recursively shortening... testing query: "lxb th"
Recursively shortening... testing query: "lxb"
Recursively shortening... testing query: "xb"
Recursively shortening... testing query: " th"
Recursively shortening... testing query: "e quic"
Recursively shortening... testing query: "k brown fox "
[u'lxb']
Testing 1 of 3: omitting character l
Testing 2 of 3: omitting character x
Testing 3 of 3: omitting character b
the minimum phrase from 'lxb' is: 'lxb'
         date                   datetime  is_canonical keyword num_results  \
0  2017-12-01 2017-12-01 10:21:52.045987          True     lxb        None   
               orig_keyword    result source  test_number  
0  lxb the quick brown fox   censored   list            1  

# without recursion: sends 1 + len(keyword without spaces) = 20 requests
# with recursion: sends 1 + 7 + 3 = 11 requests
```


Tracing through your dataframe set of sample queries:

```
0. "hello" - as expected 
1. "lxb ok" - shortened to "lxb" by recursion, "lxb" confirmed to be minimum by omission, expected.  
2. "yes lxb ok" - skipped, because 2 table entries were made above and continue_interruptions=False. Unexpected
3. "nolxbok" - recursive function cleaves as: "nol"+"xbok". Since both msgs are uncensored, script reverts to test of full query by char omission to find "lxb" as censored word. Expected
4. "lxb" - skipped. 2 table entries made above. Unexpected
5. u"习胞子" - "no_results", not "censored" so did not test for canonical keyword, expected.
```

Changes made:
 - corrected sleep to sleep_secs
 - fixed the bug above where continue_interruptions resulted in skipped queries
 - adds printing to recursive function under verbose='all'
 - skips checks for spaces under character omission (performance improvement)



Reran your test case below, output as expected for all. 
```
Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
[GCC 5.4.0 20160609] on linux2
from blockedonweibo import weibo
import pandas as pd
sqlite_file = 'results.sqlite'
weibo.create_database(sqlite_file, overwrite=True)
cookie = weibo.load_cookies()
print(weibo.verify_cookies_work(cookie))
True
sample_keywords_df = pd.DataFrame(
    [{'keyword':'hello','source':'my dataframe'},
     {'keyword':'lxb ok','source':'my dataframe'},
     {'keyword':'yes lxb ok','source':'my dataframe'},
     {'keyword':'nolxbok','source':'my dataframe'},
     {'keyword':'lxb','source':'my dataframe'},
     {'keyword':u'习胞子','source':'my dataframe'}   ])
weibo.run(sample_keywords_df,insert=True,return_df=True,get_canonical=True,sqlite_file=sqlite_file,cookies=cookie)
(0, u'hello', 'has_results')
(1, u'lxb ok', 'censored')
Found censored search phrase; determining canonical censored keyword set
Recursively shortening... testing query: "lxb"
Recursively shortening... testing query: "xb"
Recursively shortening... testing query: " ok"
[u'lxb']
Testing 1 of 3: omitting character l
Testing 2 of 3: omitting character x
Testing 3 of 3: omitting character b
the minimum phrase from 'lxb' is: 'lxb'
(2, u'yes lxb ok', 'censored')
Found censored search phrase; determining canonical censored keyword set
Recursively shortening... testing query: "yes l"
Recursively shortening... testing query: "xb ok"
[u'yes lxb ok']
Testing 1 of 10: omitting character y
Testing 2 of 10: omitting character e
Testing 3 of 10: omitting character s
Testing 5 of 10: omitting character l
Testing 6 of 10: omitting character x
Testing 7 of 10: omitting character b
Testing 9 of 10: omitting character o
Testing 10 of 10: omitting character k
the minimum phrase from 'yes lxb ok' is: 'lxb'
(3, u'nolxbok', 'censored')
Found censored search phrase; determining canonical censored keyword set
Recursively shortening... testing query: "nol"
Recursively shortening... testing query: "xbok"
[u'nolxbok']
Testing 1 of 7: omitting character n
Testing 2 of 7: omitting character o
Testing 3 of 7: omitting character l
Testing 4 of 7: omitting character x
Testing 5 of 7: omitting character b
Testing 6 of 7: omitting character o
Testing 7 of 7: omitting character k
the minimum phrase from 'nolxbok' is: 'lxb'
(4, u'lxb', 'censored')
Found censored search phrase; determining canonical censored keyword set
Recursively shortening... testing query: "xb"
[u'lxb']
Testing 1 of 3: omitting character l
Testing 2 of 3: omitting character x
Testing 3 of 3: omitting character b
the minimum phrase from 'lxb' is: 'lxb'
/.../venv/local/lib/python2.7/site-packages/pandas/core/algorithms.py:410: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
  f = lambda x, y: htable.ismember_object(x, values)
(5, u'\u4e60\u80de\u5b50', 'no_results')
         date                   datetime  is_canonical keyword num_results  \
0  2017-12-01 2017-12-01 10:03:16.879631         False   hello    76753826   
0  2017-12-01 2017-12-01 10:05:52.890120          True     lxb        None   
0  2017-12-01 2017-12-01 10:09:53.509698          True     lxb        None   
0  2017-12-01 2017-12-01 10:13:18.274100          True     lxb        None   
0  2017-12-01 2017-12-01 10:15:05.976933          True     lxb        None   
0  2017-12-01 2017-12-01 10:15:25.155217         False     习胞子        None   
  orig_keyword       result        source  test_number  
0         None  has_results  my dataframe            1  
0       lxb ok     censored  my dataframe            1  
0   yes lxb ok     censored  my dataframe            1  
0      nolxbok     censored  my dataframe            1  
0          lxb     censored  my dataframe            1  
0         None   no_results  my dataframe            1  

```

jasonqng commented:
Makes sense, glad the comments were helpful!
